state_machine/states/impl/waypoint_impl.py

```diff
@@ -52,98 +52,111 @@ async def run(self: Waypoint) -> State:
-                    line_segment.p_2.y,
-                    boundary_points[0].zone_number,
-                    boundary_points[0].zone_letter,
+        goto_points: list[Point]
+        try:
+            goto_points = list(
+                pathfinding.shortest_path_between(
+                    Point(drone_easting, drone_northing),
+                    Point(waypoint.easting, waypoint.northing),
+                    search_graph,
                 )
+            )
+        except RuntimeError:
+            # Unable to find path that doesn't intersect boundary
+            # This should never happen
+            goto_points = [Point(waypoint.easting, waypoint.northing)]
+
+        path_length: float = (
+            sum(
+                line_segment.length()
+                for line_segment in LineSegment.from_points(goto_points, False)
+            )
+            + (goto_points[0] - Point(drone_easting, drone_northing)).distance_from_origin()
+        )
 
-                # Gradually move toward goal altitude
-                curr_altitude += altitude_slope * line_segment.length()
+        curr_altitude: float = drone_position.relative_altitude_m
+        altitude_slope: float = (waypoint.altitude - curr_altitude) / path_length
 
-                try:
-                    await move_to(self.drone.system, lat_deg, lon_deg, curr_altitude, 0.83)
-                except ActionError:
-                    logging.warning(ActionError)
+        goto_points.pop()  # The last point is just the waypoint
 
-            # use 0.9 for fast_param to get within 25 ft of waypoint with plenty of leeway
-            # while being fast (values above 5/6 and less than 1 check for lat and lon with
-            # 5 digit of precision, or about 1.11 m)
-            await move_to(self.drone.system, lat_deg, lon_deg, waypoint.altitude, 0.9)
+        lat_deg: float
+        lon_deg: float
+        lat_deg, lon_deg = utm.to_latlon(
+            waypoint.easting, waypoint.northing, waypoint.zone_number, waypoint.zone_letter
+        )
 
-            logging.info("Reached waypoint %d", waypoint_num)
+        logging.info("Moving to waypoint %d (lat=%f, lon=%f)", waypoint_num, lat_deg, lon_deg)
 
-        return (ODLC if self.drone.odlc_scan else Airdrop)(self.drone, self.flight_settings)
+        for line_segment in LineSegment.from_points(goto_points, False):
+            lat_deg, lon_deg = utm.to_latlon(
+                line_segment.p_2.x,
+                line_segment.p_2.y,
+                boundary_points[0].zone_number,
+                boundary_points[0].zone_letter,
+            )
 
-    except asyncio.CancelledError as ex:
-        logging.error("Waypoint state canceled")
-        raise ex
-    finally:
-        pass
+            # Gradually move toward goal altitude
+            curr_altitude += altitude_slope * line_segment.length()
+
+            try:
+                await move_to(self.drone.system, lat_deg, lon_deg, curr_altitude, 0.83)
```

Comment:
The fast param here, was this verified to get us within 25ft without the long GPS verification time?

Reply:
No. It should be 5 / 6 (`int(6 * 0.83) == 4` and `int(6 * (5 / 6)) == 5`). However, this move_to() is for intermediate points in the path to the next waypoint, though losing too much precision could cause the drone to move out of bounds even with the 5-meter boundary shrinkage applied before pathfinding.